### PR TITLE
[WIP] Prototype of invoke_simd API and necessary dependencies.

### PR DESCRIPTION
**Pull request by kbobrovs**
This is an initial prototype of invoke_simd to get going. See comments in files for details on
std::experimental::simd stub implementation (90% - copy-paste from clang's libcxx).
The single test can run under lit.

@rolandschulz, @Pennycook, @iburyl - feel free to commit to this branch.
I can create a branch in intel/llvm instead, if more convenient. Let me know.

Major TODOs:
- solve SPIR-V calling convention problem (conversion of simd objects to sret/byval pointers)
- support callables other than function pointer (in such a way that IR is friendly for tracking function pointer to concrete function)
- implement lower-level intrinsics and the pass that converts __builtin_invoke_simd to those intrinsics
  (needed to support other callables, which generate complex data flow for function pointer. maybe __builtin_invoke_simd can be reused too)
- support bool arg translation
- more rigorous error checking (including subgroup size)
- implement oneapi::simd <-> esimd::simd conversion
- tests

Contributions from Roland Schulz (@rolandschulz).

Signed-off-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

### CODE REVIEW

File: sycl/include/std/experimental/simd.hpp [hunk 1]
```diff
@@ -0,0 +1,1740 @@
...
+#define _LIBCPP_COMPILER_CLANG_BASED 1
+#define _LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER 1
+#define _LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL_SIMD namespace std::experimental {
+#define _LIBCPP_END_NAMESPACE_EXPERIMENTAL_SIMD }
+#define _LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL_SIMD_ABI namespace std::experimental::simd_abi {
+#define _LIBCPP_END_NAMESPACE_EXPERIMENTAL_SIMD_ABI }
+#define _LIBCPP_INLINE_VAR inline
+#define _LIBCPP_PUSH_MACROS
+#define _LIBCPP_POP_MACROS
+#define _LIBCPP_NATIVE_SIMD_WIDTH_IN_BYTES 512 // is not really used for now for sycl::ext::oneapi::experimental::invoke_simd
```
Comment (Pennycook) [inline sycl/include/std/experimental/simd.hpp:678]: This should eventually be device-specific, right?  Do we need this long-term, and if so do we have any plans for how to make it device-specific?

File: sycl/include/sycl/ext/oneapi/experimental/uniform.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+namespace oneapi {
+namespace experimental {
+
+// TODO:
+// An implementation must issue a diagnostic if the
+// sycl::ext::oneapi::experimental::uniform class template is instantiated with
+// any of the following types:
+// sycl::item, sycl::nd_item, sycl::h_item, sycl::group, sycl::sub_group and
+// sycl::nd_range.
+template <class T> class uniform {
```
Comment (Pennycook) [inline sycl/include/sycl/ext/oneapi/experimental/uniform.hpp:25]: We had feedback from the oneAPI TAB that nothing in this definition makes it obvious that the `uniform` is immutable.  Making a note here that we should see if we can figure out a way to fix this (and update the extension specification) before this gets merged.

File: sycl/include/std/experimental/simd.hpp (hunk 1 above)
Comment (kbobrovs) [reply sycl/include/std/experimental/simd.hpp:678]: Yes, I think so. This is used to define the `simd_abi::native`, which is different for different devices.
`-target XXX` clang option where XXX != spir64 should define `_LIBCPP_NATIVE_SIMD_WIDTH_IN_BYTES` according to XXX 
target capabilities. This is also true for device-specific SYCL AOT targets (e.g. `spir64_gen`).
But for device-neutral `spir64` there is no perfect choice. There will have to be some reasonable default, but I'd expect 
SYCL people won't use it. Instead, they would use `oneapi::simd_abi::native_fixed_size` with proper kernel versioning based 
on [[sycl::requires]].

I think in oneapi we could introduce another kind of native ABI templated by a SYCL aspect corresponding to a device architecture (which will soon appear as an extension, I believe).

File: sycl/include/sycl/ext/oneapi/experimental/uniform.hpp (hunk 2 above)
Comment (kbobrovs) [reply sycl/include/sycl/ext/oneapi/experimental/uniform.hpp:25]: Thanks for the note. Whenever spec is updated we can update the code.

**Comment by kbobrovs** [comment]
the POC is now being productized via other PRs, this one is no longer needed

**Pull request closed without merging**